language/tools/move-package/src/package_hooks.rs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::source_package::parsed_manifest::NodeInfo;
+use anyhow::bail;
+use move_symbol_pool::Symbol;
+use once_cell::sync::Lazy;
+use std::sync::Mutex;
+
+/// A trait providing hooks to customize the package system for a particular Move application.
+/// An instance of the trait can be registered globally.
+pub trait PackageHooks {
+    /// Returns custom fields allowed in `PackageInfo`.
+    fn custom_package_info_fields(&self) -> Vec<String>;
+
+    /// A resolver for `node` dependencies in the manifest. This is called to download the
+    /// dependency from the node into the `info.local_path` location, similar as with git
+    /// dependencies.
+    fn resolve_node_dependency(&self, dep_name: Symbol, info: &NodeInfo) -> anyhow::Result<()>;
+}
+static HOOKS: Lazy<Mutex<Option<Box<dyn PackageHooks + Send + Sync>>>> =
```

Comment:
Could I ask why we're implementing this as a global instead of passing in the resolver as an argument? To avoid breaking the existing API?

Reply:
Please see the PR description. Its a large refactoring touching move-cli and move-packages. 

- Pass it as `<X>` you now have the type parameter everywhere (going deep down into static functions)
- Pass it as `dyn X` you can attach this do any of the existing context data structures because they need to be serializable/copyable etc. which doesn't work with traits.

This thing is not well designed if it comes to compositionally and extensibility. Really, any reasonable tool should have some global `Context`, `Env` whatever which is uniformly used everywhere, and does have bounds so you can add arbitrary fields. The move model has it, the vm, and it was added to the compiler sometime last year. 

Getting away without this global context is pure functional programming ideology, and it doesn't work in practice as we are seeing here once again.

Reply:
> Pass it as dyn X you can't attach this to any of the existing context data structures because they need to be serializable/copyable etc. which doesn't work with traits.

I think you can make `Clone, Serialize etc.` super traits, no? Anyway I feel like this is an orthogonal issue you'll have to deal with no matter you use a global context or manually pass it around. 

> Really, any reasonable tool should have some global Context, Env whatever which is uniformly used everywhere, and does have no bounds so you can add arbitrary fields. The move model has it, the vm, and it was added to the compiler sometime last year.

> Getting away without this global context is pure functional programming ideology, and it doesn't work in practice as we are seeing here once again.

Yeah I have fixed feelings about this. On one hand I love the fact that passing the global context around manually makes it easy to reason about side effects, but on the other hand I also I agree that it can get quite annoying, especially when you have a complex call graph. I must also admit that Rust has never been "global-free" in the first place, with its usage of the global allocator and stdio. Some popular logging crates also provide a rather smooth user experience using global variables. To this point this actually feels like a philosophical question to me and I guess I really need to study this more ; )

Anyway, I think I'm fine with the current design, and I actually think it is the right thing to do to allow us to get the functionalities we need first, even in the case we eventually decide to switch to the other way around.

Reply:
With the lack of some global env/context for the package system, would it be enough to stick this thing in the `ResolutionGraph` for now? where it would be passed in during `ResolutionGraph::new`

My only real worry with sticking this in `static` (the `static HOOKS` if I'm following this closely) is discoverability. Discoverability in the sense that it is a bit of a gotcha to know it exists, a global env or anything similar is always nice and in your face when you are constructing this tool, but obviously out of scope for this PR.

Reply:
@tnowacki that is a good point so I tried to stick this into the `ResolutionGraph`. It worked to some extend (need to remove a number of `Clone` derives) but it failed with the manifest parser which is just a bunch of static functions.

I'm wondering whether the static hooks are actually the cleaner solution anyway instead of bloating up the code. The thing with this kind of hooks is that there are once set at program startup and then stay fixed. It's a bit like loading a code plugin, different than using a trait which can represent different things during process execution.

Reply:
I think the hooks are fine for now, a small comment saying something like "this should really be in a global env/context but the package system doesn't have one " might be helpful, but not really necessary